### PR TITLE
Fixing DSR by adding outbound nat endpoint policy

### PR DESCRIPTION
**Pull request by zawachte**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
Fixed DSR loopback connectivity for Windows
**Which issue(s) this PR fixes**:

Fixes #
76316
**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by ksubrmnn** [comment]
@madhanrm FYI. I'll review this tomorrow

File: pkg/proxy/winkernel/proxier.go [hunk 1]
```diff
@@ -895,9 +913,23 @@ func endpointsToEndpointsMap(endpoints *v1.Endpoints, hostname string, hns HostN
 					klog.Warningf("Ignoring invalid endpoint port %s with empty host", port.Name)
 					continue
 				}
-				isLocal := addr.NodeName != nil && *addr.NodeName == hostname
-				epInfo := newEndpointInfo(addr.IP, uint16(port.Port), isLocal, hns)
-				endpointsMap[svcPortName] = append(endpointsMap[svcPortName], epInfo)
+
```
Comment (ksubrmnn) [inline pkg/proxy/winkernel/proxier.go]: Add logs

File: pkg/proxy/winkernel/proxier.go [hunk 2]
```diff
@@ -895,9 +913,23 @@ func endpointsToEndpointsMap(endpoints *v1.Endpoints, hostname string, hns HostN
...
+				isAlreadyEndpoint := false
+				for _, endpointInfoList := range proxierendpointsMap {
+					for _, ep := range endpointInfoList {
+						if ep.ip == addr.IP {
+							klog.Infof("Already have reference to endpoint with ip: %v", addr.IP)
+							epInfo := ep
```
Comment (ksubrmnn) [inline pkg/proxy/winkernel/proxier.go]: You can just use ep directly

File: pkg/proxy/winkernel/proxier.go [hunk 3]
```diff
@@ -895,9 +913,23 @@ func endpointsToEndpointsMap(endpoints *v1.Endpoints, hostname string, hns HostN
...
+							endpointsMap[svcPortName] = append(endpointsMap[svcPortName], epInfo)
+							isAlreadyEndpoint = true
```
Comment (ksubrmnn) [inline pkg/proxy/winkernel/proxier.go]: nit: move this to beginning of block

File: pkg/proxy/winkernel/proxier.go [hunk 4]
```diff
@@ -1016,15 +1077,42 @@ func (proxier *Proxier) syncProxyRules() {
...
+					for _, po := range newHnsEndpoint.policies {
+						if po.Type == "OutBoundNAT" {
+							var outputSettings outboundnatpolicysetting
+							if err := json.Unmarshal([]byte(po.Settings), &outputSettings); err != nil {
+								klog.Errorf("Error Unmarshaling Endpoint Policy err: %v settings: %v", err, po.Settings)
+								continue
+							}
+							for _, des := range outputSettings.Destinations {
+								if des == newHnsEndpoint.ip {
+									addDSRPolicy = true
```
Comment (ksubrmnn) [inline pkg/proxy/winkernel/proxier.go]: this should remain false. addDSRPolicy implies that you need to add the policy

File: pkg/proxy/winkernel/proxier.go [hunk 5]
```diff
@@ -1016,15 +1077,42 @@ func (proxier *Proxier) syncProxyRules() {
...
+								}
+							}
+						}
+					}
+					if !addDSRPolicy {
```
Comment (ksubrmnn) [inline pkg/proxy/winkernel/proxier.go]: this should be addDSRPolicy, not the opposite since you go and add the policy

**Review by ksubrmnn — CHANGES_REQUESTED** [review]

File: pkg/proxy/winkernel/proxier.go (hunk 4 above)
Comment (zawachte-msft) [reply pkg/proxy/winkernel/proxier.go]: Changing to "hasDSRPolicy". Bad form to set a flag to true.

**Comment by ksubrmnn** [comment]
@Microsoft/containernetworking-msft

**Comment by ksubrmnn** [comment]
@Microsoft/containernetworking-msft

File: pkg/proxy/winkernel/proxier.go (hunk 4 above)
Comment (feiskyer) [reply pkg/proxy/winkernel/proxier.go]: this is still not fixed?

**Review by feiskyer — COMMENTED** [review]
LGTM in general. Please fix lint errors and addressed all comments.

/kind bug
/priority important-soon
/approve

**Comment by feiskyer** [comment]
Please also add a release note for the change.

**Comment by ksubrmnn** [comment]
Please also add a test to verify that refcount is updated appropriately when an endpoint is shared by two services

File: pkg/proxy/winkernel/proxier.go [hunk 6]
```diff
@@ -932,6 +964,35 @@ func serviceToServiceMap(service *v1.Service, hns HostNetworkService) proxyServi
 	return serviceMap
 }
 
+func addDSRPolicyToEndpoint(hns HostNetworkService, endpoint *endpointsInfo) (*policiesinfo, error) {
```
Comment (ksubrmnn) [inline pkg/proxy/winkernel/proxier.go]: endpointsInfo already has an hns member so no need to pass it in here

**Comment by madhanrm** [comment]
Are we proceeding with this PR? can we cancel this and close since this is taken care of in CNI.

**Pull request closed without merging**